diffrax/_step_size_controller/jump_step_wrapper.py

```diff
@@ -0,0 +1,410 @@
+            inner_state,
+        )
+
+        return t1, state
+
+    def adapt_step_size(
+        self,
+        t0: RealScalarLike,
+        t1: RealScalarLike,
+        y0: Y,
+        y1_candidate: Y,
+        args: Args,
+        y_error: Optional[Y],
+        error_order: RealScalarLike,
+        controller_state: _JumpStepState[_ControllerState],
+    ) -> tuple[
+        BoolScalarLike,
+        RealScalarLike,
+        RealScalarLike,
+        BoolScalarLike,
+        _JumpStepState[_ControllerState],
+        RESULTS,
+    ]:
+        (
+            made_jump,
+            prev_dt,
+            i_step,
+            i_jump,
+            i_rjct,
+            rjct_buff,
+            step_ts,
+            jump_ts,
+            inner_state,
+        ) = controller_state.get()
+
+        # Let the controller do its thing
+        (
+            keep_step,
+            next_t0,
+            next_t1,
+            _,
+            inner_state,
+            result,
+        ) = self.controller.adapt_step_size(
+            t0, t1, y0, y1_candidate, args, y_error, error_order, inner_state
+        )
+
+        # This is just a logging utility for testing purposes
+        if self.callback_on_reject is not None:
+            jax.debug.callback(self.callback_on_reject, keep_step, t1)
+
+        # Check whether we stepped over an element of step_ts or jump_ts or rjct_buff
+        # This is all still bookkeeping for the PREVIOUS STEP.
+        if step_ts is not None:
+            # If we stepped to `t1 == step_ts[i_step]` and kept the step, then we
+            # increment i_step and move on to the next t in step_ts.
+            step_inc_cond = keep_step & (t1 == _get_t(i_step, step_ts))
+            i_step = jnp.where(step_inc_cond, i_step + 1, i_step)
+
+        if jump_ts is not None:
+            next_jump_t = _get_t(i_jump, jump_ts)
+            jump_inc_cond = keep_step & (t1 >= eqxi.prevbefore(next_jump_t))
+            i_jump = jnp.where(jump_inc_cond, i_jump + 1, i_jump)
+
+        if self.rejected_step_buffer_len > 0:
+            assert rjct_buff is not None
+            # If the step ended at t1==rjct_buff[i_rjct], then we have successfully
+            # stepped to this time and we increment i_rjct.
+            # We increment i_rjct even if the step was rejected, because we will
+            # re-add the rejected time to the buffer immediately.
+            rjct_inc_cond = t1 == _get_t(i_rjct, rjct_buff)
+            i_rjct = jnp.where(rjct_inc_cond, i_rjct + 1, i_rjct)
+
+            # If the step was rejected, then we need to store the rejected time in the
+            # rejected buffer and decrement the rejected index.
+            i_rjct = jnp.where(keep_step, i_rjct, i_rjct - 1)
+            i_rjct = eqx.error_if(
+                i_rjct,
+                i_rjct < 0,
+                "Maximum number of rejected steps reached. "
+                "Consider increasing JumpStepWrapper.rejected_step_buffer_len.",
+            )
```

Comment:
And in particular we should be able to skip this check if we set this value to `max_steps`.

Reply:
I'll run some timing tests to see if a long buffer significantly hinders performance. If not, I'm happy to change it to `max_steps`.

Reply:
So it seems that using a long buffer doesn't impact performance at all. So I think it could be fine to just make a binary parameter `revisit_rejected` and if it is `True`, then the buffer length is `max_steps`. However when vmaping over this that could create a significant but unnecessary memory overhead, so maybe giving the option to change this buffer length might be reasonable. What do you think?

Reply:
Ah, good point about the vmap case.

I think the parameter `revisit_rejected` could be of type `bool | int`. If `False` then don't revisit. If `True` then do revisit and use a buffer of size `max_steps`. If an integer then do revisit and use a buffer of that size. Gives an easy choice whilst also offering control if that's really necessary for some reason.